File: src/Bot.ts
import { Bot as BotInterface, BotStatus, Order } from './types';

// Processing time in milliseconds (10 seconds)
const PROCESSING_TIME_MS = 10000;

// Callback types
export type OnOrderCompleteCallback = (botId: number, order: Order) => void;
export type OnBotIdleCallback = (botId: number) => void;

/**
 * Bot class that processes orders
 */
export class Bot {
  private bot: BotInterface;
  private onOrderComplete: OnOrderCompleteCallback | null = null;
  private onBotIdle: OnBotIdleCallback | null = null;

  constructor(id: number) {
    this.bot = {
      id,
      status: BotStatus.IDLE,
      currentOrder: null,
      processingTimer: null
    };
  }

  get id(): number {
    return this.bot.id;
  }

  get status(): BotStatus {
    return this.bot.status;
  }

  get currentOrder(): Order | null {
    return this.bot.currentOrder;
  }

  get isIdle(): boolean {
    return this.bot.status === BotStatus.IDLE && this.bot.currentOrder === null;
  }

  get isProcessing(): boolean {
    return this.bot.status === BotStatus.ACTIVE && this.bot.currentOrder !== null;
  }

  /**
   * Set callback for when order completes
   */
  setOnOrderComplete(callback: OnOrderCompleteCallback): void {
    this.onOrderComplete = callback;
  }

  /**
   * Set callback for when bot becomes idle
   */
  setOnBotIdle(callback: OnBotIdleCallback): void {
    this.onBotIdle = callback;
  }

  /**
   * Start processing an order
   */
  startProcessing(order: Order): void {
    if (this.bot.currentOrder !== null) {
      throw new Error(`Bot ${this.bot.id} is already processing an order`);
    }

    this.bot.currentOrder = order;
    this.bot.status = BotStatus.ACTIVE;
    order.processingStartedAt = new Date();
    order.processedByBotId = this.bot.id;

    // Set timer for order completion
    this.bot.processingTimer = setTimeout(() => {
      this.completeOrder();
    }, PROCESSING_TIME_MS);
  }

  /**
   * Complete the current order
   */
  private completeOrder(): void {
    if (this.bot.currentOrder === null) return;

    const completedOrder = this.bot.currentOrder;
    completedOrder.completedAt = new Date();

    this.bot.currentOrder = null;
    this.bot.processingTimer = null;
    this.bot.status = BotStatus.IDLE;

    // Notify callback
    if (this.onOrderComplete) {
      this.onOrderComplete(this.bot.id, completedOrder);
    }

    // Notify that bot is now idle and ready for next order
    if (this.onBotIdle) {
      this.onBotIdle(this.bot.id);
    }
  }

  /**
   * Stop processing and return the current order
   * Called when bot is being destroyed
   */
  stopProcessing(): Order | null {
    if (this.bot.processingTimer) {
      clearTimeout(this.bot.processingTimer);
      this.bot.processingTimer = null;
    }

    const order = this.bot.currentOrder;
    this.bot.currentOrder = null;
    this.bot.status = BotStatus.IDLE;

    return order;
  }

  /**
   * Destroy the bot (cleanup)
   */
  destroy(): void {
    this.stopProcessing();
  }
}

File: internal/domain/ordercontroller/bot.go
package ordercontroller

// BotState 表示烹饪 Bot 的运行状态。
type BotState string

const (
	BotStateIdle       BotState = "IDLE"       // 空闲，未绑定订单
	BotStateProcessing BotState = "PROCESSING" // 正在处理订单
)

// Bot 是烹饪机器人实体，同一时刻最多处理一个订单。
type Bot struct {
	ID           int      // Bot 编号，按创建顺序从 1 递增
	State        BotState // 当前状态
	CurrentOrder *Order   // 正在处理的订单；空闲时为 nil
	PickupIndex  int      // 取单时订单在逻辑待处理队列中的 0-based 索引，用于 -bot 回插
}

// NewBot 创建指定 ID 的空闲 Bot。
func NewBot(id int) Bot {
	return Bot{ID: id, State: BotStateIdle, PickupIndex: -1}
}

// IsIdle 判断 Bot 是否处于空闲状态。
func (b *Bot) IsIdle() bool {
	return b.State == BotStateIdle
}

// IsProcessing 判断 Bot 是否正在处理订单。
func (b *Bot) IsProcessing() bool {
	return b.State == BotStateProcessing && b.CurrentOrder != nil
}

// CanAcceptOrder 判断 Bot 是否可以接受新订单（空闲或刚清空处理状态）。
func (b *Bot) CanAcceptOrder() bool {
	return b.State == BotStateIdle || (b.State == BotStateProcessing && b.CurrentOrder == nil)
}

// StartProcessing 让 Bot 开始处理订单，并记录取单时的队列位置。
func (b *Bot) StartProcessing(order Order, pickupIndex int) {
	b.State = BotStateProcessing
	b.CurrentOrder = &order
	b.PickupIndex = pickupIndex
}

// ClearProcessing 清除当前处理中的订单信息，Bot 仍保持 PROCESSING 状态以便链式取单。
func (b *Bot) ClearProcessing() {
	b.CurrentOrder = nil
	b.PickupIndex = -1
}

// SetIdle 将 Bot 设为空闲并清除所有处理状态。
func (b *Bot) SetIdle() {
	b.State = BotStateIdle
	b.ClearProcessing()
}
